clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/consumer/AsyncHandlingAutoCommitKafkaConsumer.java

```diff
@@ -0,0 +1,449 @@
+
+        /**
+         * Creates a new TopicPartitionOffsets object.
+         *
+         * @param topicPartition The topic partition to store the offsets for.
+         * @throws NullPointerException if topicPartition is {@code null}.
+         */
+        TopicPartitionOffsets(final TopicPartition topicPartition) {
+            this.topicPartition = Objects.requireNonNull(topicPartition);
+        }
+
+        /**
+         * Registers the given offset. To be invoked when the record has just been received.
+         * @param offset The offset to register.
+         */
+        public OffsetsQueueEntry setRecordReceived(final long offset) {
+            cleanupAndUpdateLastCompletedOffset();
+            final OffsetsQueueEntry queueEntry = new OffsetsQueueEntry(offset);
+            queue.add(queueEntry);
+            return queueEntry;
+        }
+
+        /**
+         * Gets the partition offset to use for an offset commit. That is the offset of the last record in the row of
+         * fully handled records if it either hasn't been committed yet or the time of the last commit is older than
+         * the 'skipOffsetRecommitPeriodSeconds'. Note that for the actual commit, {@code 1} has to be added to the
+         * returned value.
+         * <p>
+         * Otherwise an empty Optional is returned.
+         *
+         * @return The offset wrapped in an Optional or an empty Optional if no offset commit is needed.
+         */
+        public Optional<Long> getLastSequentiallyCompletedOffsetForCommit() {
+            cleanupAndUpdateLastCompletedOffset();
+            if (lastSequentiallyCompletedOffset == -1) {
+                return Optional.empty();
+            }
+            if (!queue.isEmpty()) {
+                log.trace("getOffsetsToCommit: record with offset {} to use for commit is {} entries behind last received offset {}; partition [{}]",
+                        lastSequentiallyCompletedOffset, queue.size(), queue.getLast().getOffset(), topicPartition);
+            }
+            if (lastSequentiallyCompletedOffset != lastCommittedOffset) {
+                return Optional.of(lastSequentiallyCompletedOffset);
+            } else if (lastCommitTime != null
+                    && lastCommitTime.isBefore(Instant.now().minusSeconds(skipOffsetRecommitPeriodSeconds))) {
+                log.trace("getOffsetsToCommit: record with offset {} will be recommitted (last commit {} too long ago); partition [{}]",
+                        lastSequentiallyCompletedOffset, lastCommitTime, topicPartition);
+                return Optional.of(lastSequentiallyCompletedOffset);
+            } else {
+                return Optional.empty();
+            }
+        }
+
+        private void cleanupAndUpdateLastCompletedOffset() {
+            while (Optional.ofNullable(queue.peek()).map(OffsetsQueueEntry::isHandlingComplete).orElse(false)) {
+                lastSequentiallyCompletedOffset = queue.remove().getOffset();
```

Comment:
`cleanupAndUpdateLastCompletedOffset()` is invoked during several independent events (eg. commit at regular intervals, arrival of a new record and during re-balancing).
1.  In order to ensure that the correct item is being removed I think it's better to use `queue.remove(offsetsQueueEntry)` instead of invoking `queue.remove()`.
2. The operations `peek()` and `remove()` are not done atomically. Since there are several events that trigger this, wouldn't that affect the sequence of the offsets being committed. For example `cleanupAndUpdateLastCompletedOffset()` is also being invoked by a non _synchronized_ method namely `setRecordReceived(...)`

Reply:
The `cleanupAndUpdateLastCompletedOffset()` method is invoked from these methods (in the outer class):
- `setRecordReceived`
- `getOffsetsToCommit`
- `clearObsoleteTopicPartitionOffsets`
All of these are `synchronized`. And the used `OffsetsQueueEntry#handlingComplete` field is an `AtomicBoolean`. So I don't see any potential concurrency issues here.